design/src/rfcs/rfc_0020_endpoints_20.md

```diff
@@ -0,0 +1,599 @@
+struct CustomDdbResolver;
+
+impl ResolveEndpoint<aws_sdk_dynamodb::endpoint::Params> for CustomDdbResolver {
+    fn resolve_endpoint(&self, params: &Params) -> Result<Endpoint, EndpointResolutionError> {
+        // custom resolver to redirect to DDB local if a flag is set
+        let base_endpoint = aws_sdk_dynamodb::endpoint::default_resolver().resolve_endpoint(params).expect("valid endpoint should be resolved");
+        if env::var("LOCAL") == Ok("true") {
+            // update the URI on the returned endpoint to localhost while preserving the other properties
+            base_endpoint.builder().uri("http://localhost:8888").build()
+        } else {
+            base_endpoint
+        }
+    }
+}
+
+async fn main() {
+    let conf = aws_config::load_from_env().await;
+    let ddb_conf = aws_sdk_dynamodb::config::Builder::from(&conf).endpoint_resolver(CustomDdbResolver);
+    let dynamodb = aws_sdk_dynamodb::Client::from_conf(ddb_conf);
+}
+```
+
+> **What about alternative S3 implementations? How do we say "don't put prefix bucket on this?"**
+>
+> For cases where users want to use the provided URL directly with no modification users will need to rely on service
+> specific configuration, like forcing path style addressing for S3.
+
+_Alternative Design_: [Context Aware Endpoint Trait](#context-aware-endpoint-traits)
+
+> _Optional addition_: We could add an additional `EndpointResolver` parameter to `SdkConfig` that exposed a global trait
+where `Params` is `&dyn Any` similar to [Context Aware Endpoint Trait](#context-aware-endpoint-traits). If these were
+**both** set, a runtime panic would alert users to the misconfiguration.
+
+### New Endpoint Traits
+
+The new endpoint resolution trait and `Endpoint` struct will be Smithy-native. AWS endpoint middleware will pull
+the `Endpoint`
+out of the property bag and read the properties to determine auth/signing + any other AWS metadata that may be required.
+
+An example of the `Endpoint` struct is below. This struct will be in `aws-smithy-types`, however, it should initially be
+gated with documentation warning about stability.
+
+#### The Endpoint Struct
+
+```rust
+// module: `aws_smithy_types::endpoint`
+// potential optimization to reduce / remove allocations for keys which are almost always static
+// this can also just be `String`
+type MaybeStatic<T> = Cow<'static, T>;
+
+/// Endpoint
+#[derive(Debug, PartialEq)]
+pub struct Endpoint {
+    // Note that this allows `Endpoint` to contain an invalid URI. During conversion to an actual endpoint, the
+    // the middleware can fail, returning a `ConstructionFailure` to the user
+    url: MaybeStatic<str>,
+    headers: HashMap<MaybeStatic<str>, Vec<MaybeStatic<str>>>,
+    properties: HashMap<MaybeStatic<str>, aws_smithy_types::Document>,
+}
+
+// not shown:
+// - impl block with standard accessors
+// - builder, designed to be invoked / used by generated code
+```
+
+> **What's an Endpoint property?**
```

Comment:
I think this needs to be higher up since endpoint properties are mentioned in code samples above.